sigstore-java/src/main/java/dev/sigstore/http/UnsuccessfulResponseHandler.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2023 The Sigstore Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package dev.sigstore.http;
+
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.HttpUnsuccessfulResponseHandler;
+import com.google.api.client.util.BackOff;
+import com.google.api.client.util.BackOffUtils;
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.api.client.util.Sleeper;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.time.Instant;
+import org.apache.http.client.utils.DateUtils;
+
+/**
+ * Handler for 429 and standard server errors 5XX. Objects of this class are single use. A new
+ * instance should be created for each request (but this is handled in {@link
+ * HttpClients#newRequestFactory(HttpParams)}).
+ *
+ * <p>This implementation is a bit weird, but hopefully we can just move to grpc for rekor when
+ * ready.
+ */
+public class UnsuccessfulResponseHandler implements HttpUnsuccessfulResponseHandler {
+  private final Sleeper sleeper;
+
+  // for 5XX
+  private final BackOff exponentialBackOff;
+
+  // for 429
+  private static final int MAX_429_RETRIES = 10;
+  private int current429Retries = 0;
+
+  public static UnsuccessfulResponseHandler newUnsuccessfulResponseHandler() {
+    return new UnsuccessfulResponseHandler(Sleeper.DEFAULT, new ExponentialBackOff());
+  }
+
+  @VisibleForTesting
+  UnsuccessfulResponseHandler(Sleeper sleeper, ExponentialBackOff exponentialBackOff) {
+    this.sleeper = sleeper;
+    this.exponentialBackOff = exponentialBackOff;
+  }
+
+  @Override
+  public boolean handleResponse(HttpRequest request, HttpResponse response, boolean supportsRetry)
+      throws IOException {
+    if (!supportsRetry) {
+      return false;
+    }
+    try {
+      // exponential backoff for 5XX
+      if (response.getStatusCode() / 100 == 5) {
+        return BackOffUtils.next(sleeper, exponentialBackOff);
```

Comment:
`Retry-After` can easily appear in 503, so it would probably be nice to handle it as well

Reply:
Oh right. Kk lemme work that in